tsdb/head_wal.go

```diff
@@ -228,9 +228,17 @@ Outer:
 				// It is possible that some old sample is being processed in processWALSamples that
 				// could cause race below. So we wait for the goroutine to empty input the buffer and finish
 				// processing all old samples after emptying the buffer.
+				select {
+				case <-outputs[idx]: // allow output side to drain to avoid deadlock
+				default:
+				}
```

Comment:
We're draining `output[idx]`, but not doing anything with data and instead losing it. Is this a correct fix?

Reply:
output[idx] is the slice that contains already processed samples, so it is safe to discard

Reply:
Thanks for reply. If we can safely discard values from `output[idx]`, then what's its purpose? Is it used to keep two goroutines in lockstep?

Reply:
`outputs[idx]` gives the slice for reuse instead of creating new slices everytime. See:
https://github.com/prometheus/prometheus/blob/280eaf33f0e935c18465c7aed3dcdd7a8cdfb6c4/tsdb/head_wal.go#L280-L285
And since we will face duplicate series records not so often, the above draining won't cause any performance degradation overall.

Reply:
Thanks for explanation!

Reply:
I had the same confusion; I tried to change the comment on `processWALSamples` to explain better but maybe not enough.

As an experiment I also re-coded this buffer re-use as a `sync.Pool` and it was slower so 🤷